### DISCUSSION TITLE
重试消息一次发送多次，只能通过幂等处理吗？

### DISCUSSION THREAD

**Discussion by yyuaaon** [apache/rocketmq#8131] [Q&A]
retrytimes为0时刻意抛出异常，模拟重试场景。结果重试消息监听到了2次，而且2次的重试消息retrytimes都为1。这是为什么？

**Comment by fuyou001** [comment]
是的，必须要做幂等。

**Comment by cserwen** [comment]
首先 RocketMQ 不保证仅消费一次，消费重复是可能的，最常见的如 Rebalance 导致的重复消费，针对你这种情况，提供几个排查点：
1. 通过查询重试消息，确认是否产生了两条重试消息，产生相同重试消息的原因有很多，这里不会详细列举
2. 如果只有一条重试消息，确认下重试消费期间是否有 Rebalance ，
3. 消费组下是否存在多个 clientId 相同的消费者，clientId 相同会导致 rebalance 计算错误，将一个 queue 同时分配给多个 消费者
4. 还有其他一些特殊 case 都会造成重复消费，比如依赖冲突等，这里不再列举

所以最好还是在**消费侧增加幂等逻辑**，避免重复消费对业务造成影响

**Comment by lingmeme** [comment]
业务幂等是非常有必要的，这也是兜底的一种方式。

**Comment by 3424672656** [comment]
设置retryTimes=0，模拟重试场景，实际日志打印为times=1，不知道你是不是指这个
![图片](https://github.com/user-attachments/assets/2a2d0fd8-e988-46de-8102-4b3527631c81)
